sos/plugins/__init__.py

```diff
@@ -146,34 +146,54 @@ def __repr__(self):
         """
         return self.__str(quote=True, prefix="SoSPredicate(", suffix=")")
 
-    def __nonzero__(self):
-        """Predicate evaluation hook.
-        """
-        pvalue = False
-
-        # Allow loading kernel modules?
-        pvalue |= self._owner.get_option("allow_system_changes")
+    def _eval_kmods(self):
+        if not self.kmods or self._owner.get_option('allow_system_changes'):
```

Comment:
Do we really want to hook `--allow-system-changes` option to `kmod` predicates? Can't be there a use case where:

- `--allow-system-changes` is used
- some command can be executed *only* if some kernel module is loaded - otherwise, the command can hung or misbehave either way (I just speculate here, not being aware of such command)

If we would have such a command called by sosreport, then running `sosreport --allow-system-changes` can be dangerous. Despite we would guard that command via `kmod` "properly".

I rather think that `--allow-system-changes` shall be independent on predicate evaluation (despite plugins will need a little more code). See the option name is general and expects it to be used to potentially alter the system in either way - just the current usage is that it can load some kernel modules.

Reply:
My thought here is that if a user uses `--allow-system-changes`, we (and they) don't actually care about gating commands, so it should be an auto-pass.

In the event that a command that requires a kernel module to be loaded runs when it isn't loaded _and_ it doesn't load the kernel module itself (not aware of any), my expectation would be that the command fails rather than hangs.

It would also make plugin code a bit more clunky by having the evaluation ignore the option - we'd have to manually test the predicate and then run the command without a predicate, rather than evaluation the predicate with the command request at the same time. I believe this is why even the original evaluation checked against the use of `allow-system-changes`.